### PR TITLE
Change S3Bitstore to use TransferManager to allow larger then 5gb files

### LINKED ISSUE #7674: [DS-4334] S3 bitstore limits uploads to 5GB maximum (doesn't use Multipart upload)
<i>Imported from JIRA <a href="https://jira.lyrasis.org/browse/DS-4334">[DS-4334]</a> created by tdonohue</i>
<p>Bug logged per this discussion thread: <a href="https://groups.google.com/d/msg/dspace-tech/nLlSOzI3UZk/e_l4jhKUEAAJ" class="external-link" target="_blank" rel="nofollow noopener">https://groups.google.com/d/msg/dspace-tech/nLlSOzI3UZk/e_l4jhKUEAAJ</a></p>

<p>Simply put, the current implementation of S3BitStoreService does <b>not</b> use Amazon S3's multipart upload functionality in the "put" method: <a href="https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java#L164" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/DSpace/DSpace/blob/dspace-6_x/dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java#L164</a></p>

<p>In order to use Multipart upload, this code would need to be updated to use a TransferManager, like the S3 example Java class here: <a href="https://docs.aws.amazon.com/AmazonS3/latest/dev/HLuploadFileJava.html" class="external-link" target="_blank" rel="nofollow noopener">https://docs.aws.amazon.com/AmazonS3/latest/dev/HLuploadFileJava.html</a></p>

<p><b>This means that currently the S3Bitstore only supports uploading files that are <=5GB</b>. This is because S3 only allows PUT commands of 5GB or less, unless you are using multipart upload: <a href="https://aws.amazon.com/s3/faqs/" class="external-link" target="_blank" rel="nofollow noopener">https://aws.amazon.com/s3/faqs/</a></p>

<p>Since S3 support files sizes of 5TB or more, you can still store / migrate larger files to S3 <b>manually</b>, but they cannot be uploaded via DSpace itself.</p>

<p> </p>

### PR DESCRIPTION
**Pull request by ConfusionOrb221**
## References
* Fixes DSpace/DSpace/issues/7674

## Description
Changed the S3BitstoreService put method to use the TransferManager instead of PutObjectRequest, this allows the DSpace the ability to upload larger than 5gb files.

## Instructions for Reviewers

List of changes in this PR:
* S3BitstoreService put method
* TransferManager instead of PutObjectRequest

To test this feature a couple thing will need to be done:
- Change the `spring.servlet.multipart.max-file-size` and the `spring.servlet.multipart.max-request-size` either inside of your local.cfg or application.properties file to reasonable values, locally I used -1  (unlimited) but 10G or something similar may be reasonable for your testing. This property is needed to allow the actual upload of a file this large to DSpace as the spring config default limits it to 512MB.
- The bitstore.xml file will need to be adjusted in order to use the second option `s3store` locally. I changed `<entry key="0" value-ref="localStore"/>` to `<entry key="0 value-ref="s3Store"/>` and then provided the s3Store with proper keys/credentials to a test s3 bucket.
- Then upload a file to the DSpace that is larger then 5GB I used the `mkfile 8G` command to create a test file but if you have a large file just sitting around that'll work as well.
- After this I just `ls` the aws bucket to see that the file was uploaded when it had completed and verified the size.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java
Comment (cwilper) [inline dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java:185]: This should use the CSA constant instead of hardcoding "MD5" here. That will probably get rid of the lgtm warning as well.

**Comment by Raf-atmire** [comment]
Hi @ConfusionOrb221 , Thank you for the PR!
With the help of @ybnd , this PR was tested but he revealed a tiny bit of feedback that we'd like to address.
After the file has been uploaded to REST (once progress bar reaches 100% AFAIK?) the upload remains shown w/ “Processing…” – this can take quite a long time for large files
At this point it is unclear whether it is safe to close the page or if the upload will continue if the user gets logged out after idling for too long
@ybnd  confirmed that both are fine.
A simple message like “it is safe to close this page now” would help a lot

**Comment by jonas-atmire** [comment]
A small extension of Raf's comment on the workings of this PR 

We've used this code internally for an import of a lot of items into the DSpace codebase in combination with the S3 assetstores, where some of them had quite large bitstreams (20G+)
The large size imports posed no problems, but we did encounter an issue with the number of bitstreams being uploaded.

After a while, the upload failed without an actual relevant message being logged.

```
com.amazonaws.SdkClientException: Upload canceled
        at com.amazonaws.services.s3.transfer.internal.UploadMonitor.call(UploadMonitor.java:137) ~[aws-java-sdk-s3-1.12.45.jar:?]
        at com.amazonaws.services.s3.transfer.internal.UploadMonitor.call(UploadMonitor.java:45) ~[aws-java-sdk-s3-1.12.45.jar:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
        at java.lang.Thread.run(Thread.java:829) ~[?:?]
```

It turns out that the instantiation of the transfermanager [here](https://github.com/DSpace/DSpace/pull/8060/files#diff-02dc8d6ccf99e2499bb006b204a26eb968f83223394b5b50a386d7be887ba8adR187) is done for every upload.
For regular use in the repo, that isn't apparent, but for scripts that ingest large numbers of bitstreams, this can lead to problems.

@ConfusionOrb221
To fix this, the initialising of the TransferManager can be moved up to the [init()](https://github.com/DSpace/DSpace/blob/5ae5398b99220e8cf96cc59a77a22b4151433284/dspace-api/src/main/java/org/dspace/storage/bitstore/S3BitStoreService.java#L91) method of the same class. That way we are using up a lot less resources
Afterwards, we were able to upload +40k items with bitstreams ranging from basically nothing to 25G without any issues

**Comment by ybnd** [comment]
@cwilper ↑ looks like that wasn't enough to fool LGTM
But since this is for a checksum I suppose we can just ignore it?

**Comment by tdonohue** [comment]
@ybnd and @cwilper :  Just to verify, yes, you can ignore that LGTM warning when it appears for checksums.   LGTM is very useful in finding some obvious code issues, but it can sometime come up with "false positives".  This is a false positive as we are not using MD5 for any sort of encryption

**Comment by tdonohue** [comment]
Looks like this now has merge conflicts and would need to be updated based on the improved S3 implementation work in #8356 (which was just merged)

**Comment by ybnd** [comment]
@tdonohue AFAIK this has been tested against S3 by @jonas-atmire, @jensvannerum and me for sure

**Comment by tdonohue** [comment]
@jensvannerum : This appears to be failing it's own unit tests now.

**Comment by jensvannerum** [comment]
> @jensvannerum : This appears to be failing it's own unit tests now.

This was indeed the case because the tests of PR https://github.com/DSpace/DSpace/pull/8060#:~:text=implementation%20work%20in-,%238356,-(which%20was%20just were made against the code that this PR changes. I committed my changes to the tests and all should be fine now

**Comment by abollini** [comment]
I noted this PR due to some messages on the slack channel offering support for testing, sorry to have miss it before.
We have implemented upload via transfer manager since 2020 already in DSpace(-CRIS) 5 I was convinced that some PR from us was sitting somewhere but I forgot about it.
Here some reference to the commit / code on the official, public, dspace-cris codebase
https://github.com/4Science/DSpace/commit/705a3faa5c1c8f4d0c4b891413bce64eed6afb9b
That said, we have also recently updated this code and included it in dspace-cris 7 (it has been included in the release of the 30th August 2022.01.02) 
https://github.com/4Science/DSpace/commit/030396cd1b4ea1ce8229b264b2d2b5883a5954a4
and improved later with
https://github.com/4Science/DSpace/commit/08912eaed8750d3630636a5ecf3314e7eac44121

At a first glance the solution looks very similar, I suggest that @LucaGiamminonni makes a check so that we can move this PR forward opening a subsequent PR if there are any further improvements in our code that could be beneficial

**Review by LucaGiamminonni — APPROVED** [review]
HI @ConfusionOrb221, the changes for me are ok. Nothing to report. Thank you

**Review by tdonohue — APPROVED** [review]
👍 Code looks good to me.  Thanks @jensvannerum , @ConfusionOrb221 and all the others who have contributed to this.

Since this has been approved also from the 4Science team, I'll go ahead and merge immediately.

**Pull request merged**
